### PR TITLE
Closes #25

### LINKED ISSUE #25: Create dataset loader for PubMedQA
From https://github.com/pubmedqa/pubmedqa

### PR DESCRIPTION
**Pull request by SamuelCahyawijaya**
Please name your PR after the issue it closes. You can use the following line: "Closes #ISSUE-NUMBER" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

If the following information is NOT present in the issue, please populate:

- **Name:**: PubmedQA
- **Description:**: https://pubmedqa.github.io/
- **Paper:** https://aclanthology.org/D19-1259/
- **Data:** https://github.com/pubmedqa/pubmedqa

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`. 

Note: Need to specify a specific `subset_id`, i.e., pqal, pqau, and pqaa, to run the unit test.

### CODE REVIEW

**Review by hakunanatasha — CHANGES_REQUESTED** [review]
Hi @SamuelCahyawijaya 

Please remove the images.

I'm having trouble running the unittests for `pqaa` and `pqau` but `pqal` passes - when I look at google I see it limits the download because of size - can you check if these work?

**Comment by SamuelCahyawijaya** [comment]
Hi @hakunanatasha, same as in the other PR, I have deleted the two image files here. 
For the `pqau` and `pqaa`, I think it is the bug from HF datasets as mentioned here: https://github.com/huggingface/datasets/pull/3787, and it has been included in the `datasets==2.0.0` (https://github.com/huggingface/datasets/releases/tag/2.0.0).

I tested on `datasets==2.0.0` and it works just fine. Is it possible to update the datasets requirement in the `requirements.txt` to `datasets==2.0.0` to cope with this problem?

I tested some other datasets (mediqa_qa, mediqa_rqe, pubmed_qa, paramed, pico_extraction, medhop. scital, and mqp) using the `datasets==2.0.0` and all of them seem to work just fine.

**Comment by hakunanatasha** [comment]
@SamuelCahyawijaya very interesting - yes if that's the case, let's update the reqs.

**Comment by SamuelCahyawijaya** [comment]
@hakunanatasha : I have added the 10-fold configs for source and bigbio schemas. The subset_id for `pqal` changes from `pqal_[source|bigbio]` to `pqal_fold{k}_[source|bigbio]`, k in [0..9].

**Comment by galtay** [comment]
don't worry about the images (we can just merge this and then add the images back right after) 
I'll add some comments to the code though

File: biodatasets/pubmed_qa/pubmed_qa.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+Each PubMedQA instance is composed of:
+  (1) a question which is either an existing research article title or derived from one, 
+  (2) a context which is the corresponding abstract without its conclusion, 
+  (3) a long answer, which is the conclusion of the abstract and, presumably, answers the research question, and
+  (4) a yes/no/maybe answer which summarizes the conclusion.
+
+PubMedQA is the first QA dataset where reasoning over biomedical research texts, 
+especially their quantitative contents, is required to answer the questions.
+"""
```
Comment (galtay) [inline biodatasets/pubmed_qa/pubmed_qa.py:53]: can we add more details here about what the "L", "U", and "A" subsets are please

File: biodatasets/pubmed_qa/pubmed_qa.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+class PubmedQADataset(datasets.GeneratorBasedBuilder):
+    """PubmedQA Dataset"""
+    SOURCE_VERSION = datasets.Version(_SOURCE_VERSION)
+    BIGBIO_VERSION = datasets.Version(_BIGBIO_VERSION)
+
+    BUILDER_CONFIGS = [
+        # PQAU & PQAA Source
+        BigBioConfig(
+            name="pqaa_source",
```
Comment (galtay) [inline biodatasets/pubmed_qa/pubmed_qa.py:82]: can we make the names and subset_ids consistent with the other dataset loader naming conventions? 
e.g. `pubmed_qa_a_source`

File: biodatasets/pubmed_qa/pubmed_qa.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    ] + [
+        # PQAL Source Schema        
+        BigBioConfig(
+            name=f"pqal_fold{i}_source",
+            version=datasets.Version(_SOURCE_VERSION),
+            description="PubmedQA labeled source schema",
+            schema="source",
+            subset_id=f"pqal_fold{i}"
+        ) for i in range(10)
+    ] + [
```
Comment (galtay) [inline biodatasets/pubmed_qa/pubmed_qa.py:119]: it seems that these cv splits are defined but don't change how the data is loaded below.

File: biodatasets/pubmed_qa/pubmed_qa.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+                        "filepath": data_dir / self.config.subset_id / "dev_set.json"
+                    }
+                ),
+                datasets.SplitGenerator(
+                    name=datasets.Split.TEST,
+                    gen_kwargs={
+                        "filepath": data_dir / "pqal_test_set.json"
+                    }
+                )
+            ]
```
Comment (galtay) [inline biodatasets/pubmed_qa/pubmed_qa.py:186]: if I'm reading this correctly, I think we don't need to include the cv splits

**Review by galtay — CHANGES_REQUESTED** [review]

File: biodatasets/pubmed_qa/pubmed_qa.py (hunk 4 above)
Comment (SamuelCahyawijaya) [reply biodatasets/pubmed_qa/pubmed_qa.py:186]: Hi @galtay, thanks for reviewing. The CV split is only defined for train and dev, I was thinking to merge train and validation data or using a single split before, but to strictly follow the split defined in https://github.com/pubmedqa/pubmedqa, we need to split the train and dev sets into 10-folds.

File: biodatasets/pubmed_qa/pubmed_qa.py (hunk 3 above)
Comment (SamuelCahyawijaya) [reply biodatasets/pubmed_qa/pubmed_qa.py:119]: @galtay : it will change the folder path on the `_split_generators` function for PQA-L by defining referring the folder using the subset_id.
https://github.com/bigscience-workshop/biomedical/blob/4852e8bbff8fd8b77fb583ab09ea14917f3e9645/biodatasets/pubmed_qa/pubmed_qa.py#L166-L179

File: biodatasets/pubmed_qa/pubmed_qa.py (hunk 2 above)
Comment (SamuelCahyawijaya) [reply biodatasets/pubmed_qa/pubmed_qa.py:82]: Sure, I will update it into `pubmed_qa_artificial_*`, `pubmed_qa_labelled_*`, and `pubmed_qa_unlabelled_* `

File: biodatasets/pubmed_qa/pubmed_qa.py (hunk 1 above)
Comment (SamuelCahyawijaya) [reply biodatasets/pubmed_qa/pubmed_qa.py:53]: Sure! I'll add some explanations on each subset in the header comment section.

File: biodatasets/pubmed_qa/pubmed_qa.py (hunk 3 above)
Comment (SamuelCahyawijaya) [reply biodatasets/pubmed_qa/pubmed_qa.py:119]: Oh, maybe it is because most of the train data on different folds are the same given a certain index.
I checked `fold0`, `fold1`, and `fold2`on the validation set and it seems to work as expected.

Screenshot 2022-04-15 at 11 16 35 AM

File: biodatasets/pubmed_qa/pubmed_qa.py (hunk 4 above)
Comment (galtay) [reply biodatasets/pubmed_qa/pubmed_qa.py:193]: ok, I think i see now, so there are train/dev/test splits WITHIN each fold?

**Comment by galtay** [comment]
Hi @SamuelCahyawijaya, I think I understand this dataset a bit better now.  Are all the BigBio Q/A schemas using the yes/no answers? It seems that we might also be able to support `long_answer` Q/A and maybe even text classification with the MESH labels. I think we can wrap up it up for now but let's flag this for later development.

**Review by galtay — APPROVED** [review]

File: biodatasets/pubmed_qa/pubmed_qa.py (hunk 3 above)
Comment (galtay) [reply biodatasets/pubmed_qa/pubmed_qa.py:129]: yes, I think that is what was confusing me!

**Comment by SamuelCahyawijaya** [comment]
Hi @galtay, yes, you are right. The dataset can be utilized to support `long_answer` Q/A and some other possible tasks. Before I include the long answer as the label in the bigbio schema, but then I remove it, since we need to maintain the QA type.

Let me know if later we plan to implement a different task / QA type for this. I can help to implement that part.
